### PR TITLE
architecture documentation

### PR DESCRIPTION
**Pull request by eprothro**
It took quite a bit of digging into Puma source to get the architectural sense I needed for some recent load testing / tuning tasks.

I created some internal documentation for our team that I figured I would see if y'all thought belonged here. 

Happy to discuss or change in any way, my goal is to help those wanting to understand more who don't have the time or confidence to dig in the source to understand more. It would have helped me.

### CODE REVIEW

File: docs/architecture.md [hunk 1]
```diff
@@ -0,0 +1,36 @@
+# Architecture
+
+## Overview
+
+![http://bit.ly/2iJuFky](images/puma-general-arch.png)
+
+Puma is a threaded web server, processing requests across a TCP or UNIX socket.
+
+Workers accept connections from the socket and a thread in the worker pool processes the client's request.
```
Comment (nateberkopec) [inline docs/architecture.md]: "worker pool" -> "worker's thread pool"

**Comment by nateberkopec** [comment]
One quick change there, but otherwise LGTM and thank you so much for this!

**Comment by eprothro** [comment]
@nateberkopec fixed, thanks.

**Comment by nateberkopec** [comment]
@eprothro Remind me: what did you use to make these PRs?

**Comment by eprothro** [comment]
@nateberkopec In this case I used draw.io, saved the pngs in the repo, and used a shortened link to a url that embeds the diagram content/source. You can click the diagram links from the MD file, click edit, and then make whatever changes are needed.

Draw.io now supports embedding the diagram in the png, so now I usually save diagram pngs (with embedded diagrams in them) in-repo, and future contributors have everything they need without the shortened URL...

**Comment by nateberkopec** [comment]
@eprothro nice man, thanks! 👍

**Comment by MSP-Greg** [comment]
@eprothro

> draw.io

Cool. Thanks for the link.  Exports svg also, which might be a better format for 'current' browsers...

**Comment by eprothro** [comment]
@MSP-Greg yeah, in theory I agree with you, in practice your mileage may vary with trying to embed svgs in GFMD and the github ecosystem...

**Pull request merged**